### PR TITLE
SI-40 Background-enabled Uploads

### PR DESCRIPTION
**Pull request by jellybeansoup**
In typical use of the Zone5 SDK, file uploads are often the last thing to be done before an app is backgrounded, as it typically means that someone has produced an activity recording (with the RideRecorder, natch) and is uploading it to their account on the server. Unfortunately, this often means that the upload may start, but not complete, before the user actually dismisses the app, causing the upload to be paused until the user returns to the app.

This PR routes all file upload tasks through a dedicated, background-enabled `URLSession`, which allows file uploads to continue once the app is backgrounded and/or suspended (but not terminated).

### CODE REVIEW

File: Zone5/Transport/URLRequestInterceptor.swift [hunk 1]
```diff
@@ -148,20 +149,39 @@ internal class URLRequestInterceptor: URLProtocol {
...
+				return
+			}
+
+			currentTask = session.uploadTask(with: request, fromFile: url)
+
+			var completionObserver : NSObjectProtocol? = nil
+			completionObserver = NotificationCenter.default.addObserver(forName: URLRequestDelegate.uploadCompleteNotification, object: currentTask, queue: nil) { [ weak self ] notification in
+				// remove observers.
+				if let completionObserver = completionObserver {
+					NotificationCenter.default.removeObserver(completionObserver, name: URLRequestDelegate.downloadCompleteNotification, object: currentTask)
```
Comment (jeanhall) [inline Zone5/Transport/URLRequestInterceptor.swift]: I'm guessing this is supposed to be remove uploadCompleteNotification, not downloadCompleteNotification

File: Zone5/Transport/URLRequestDelegate.swift [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		if let error = error {
+			userInfo["error"] = error
+		}
+
+		if let response = task.response {
+			userInfo["response"] = response
+		}
+
+		if let uploadTask = task as? URLSessionUploadTask {
+			if let mutableData = receivedData[uploadTask] {
```
Comment (jeanhall) [inline Zone5/Transport/URLRequestDelegate.swift]: when do you clear this map entry?

**Review by jeanhall — CHANGES_REQUESTED** [review]
Looks really good. Very clean and easy to follow. Just 2 things that need attention.

File: Zone5/Transport/URLRequestDelegate.swift (hunk 2 above)
Comment (jellybeansoup) [reply Zone5/Transport/URLRequestDelegate.swift]: Good point. I'll swap this out with a `removeValue(forKey:)` call.

File: Zone5/Transport/URLRequestInterceptor.swift (hunk 1 above)
Comment (jellybeansoup) [reply Zone5/Transport/URLRequestInterceptor.swift]: It sure is!

**Review by jeanhall — APPROVED** [review]

**Pull request merged**